### PR TITLE
NcursesDisplay.menu: treat ESC as cancel

### PR DESCRIPTION
**Pull request by pde**
Currently it will fire a weird traceback like:

```
File "/home/ubuntu/letsencrypt/letsencrypt/plugins/selection.py", line 113, in choose_plugin
    code, index = disp.menu(question, opts, help_label="More Info")
File "/home/ubuntu/letsencrypt/letsencrypt/display/util.py", line 129, in menu
    return code, int(index) - 1
ValueError: invalid literal for int() with base 10: ''
```

### CODE REVIEW

**Comment by bmw** [comment]
In general, I think the best approach here is to have all of the `NCursesDisplay` `IDisplay` methods to check whether dialog's exit code is `OK` or `HELP`. If not, it should return `CANCEL` as the exit code.

**Comment by bmw** [comment]
This hasn't seen any progress in quite a while.

Closing the PR to help with my OCD, but feel free to open again when it's ready for another review.

**Comment by pde** [comment]
Reopening this to work on it

**Comment by pde** [comment]
Reopening for real now.

File: certbot/display/util.py [hunk 1]
```diff
@@ -29,7 +29,6 @@
 HELP = "help"
 """Display exit code when for when the user requests more help."""
 
-
```
Comment (bmw) [inline certbot/display/util.py]: nit: This does nothing but cause `pep8` errors.

File: certbot/display/util.py [hunk 2]
```diff
@@ -51,6 +50,21 @@ def _wrap_lines(msg):
 
     return os.linesep.join(fixed_l)
 
+
+def _clean(dialog_result):
+    """Work around inconsistent return codes from python-dialog.
```
Comment (bmw) [inline certbot/display/util.py]: nit: The codes aren't really inconsistent. They're all [documented](http://pythondialog.sourceforge.net/doc/Dialog_class_overview.html), it's just we only use and expect a subset of them. Maybe this comment should reflect that more accurately?

File: certbot/display/util.py [hunk 3]
```diff
@@ -126,9 +140,9 @@ def menu(self, message, choices, ok_label="OK", cancel_label="Cancel",
                 (str(i), choice) for i, choice in enumerate(choices, 1)
             ]
             # pylint: disable=star-args
-            code, index = self.dialog.menu(message, **menu_options)
+            code, index = _clean(self.dialog.menu(message, **menu_options))
 
-            if code == CANCEL:
+            if code == CANCEL or index == "":
```
Comment (bmw) [inline certbot/display/util.py]: Question: Is this necessary after checking the exit code? Is it possible to have a OK or HELP exit code and an empty index string?

File: certbot/display/util.py [hunk 4]
```diff
@@ -164,6 +178,7 @@ def yesno(self, message, yes_label="Yes", no_label="No", **unused_kwargs):
         :rtype: bool
 
         """
+        assert OK == self.dialog.DIALOG_OK, "What kind of absurdity is this?"
```
Comment (bmw) [inline certbot/display/util.py]: nit: If we want to leave this in, maybe we should check it at the top of the file or in `__init__` to catch the problem if we actually think there might be one early.

**Comment by bmw** [comment]
What about the `checklist` and `directory_select` methods?

File: certbot/display/util.py (hunk 3 above)
Comment (pde) [inline certbot/display/util.py]: I can't remember why I added this now.  Perhaps I saw this for real, or perhaps I was just being defensive?

File: certbot/display/util.py (hunk 4 above)
Comment (pde) [inline certbot/display/util.py]: The reason this is here is the same as the reason we have these janky constants in the first place: they're not available from the `dialog` module, only from `dialog.Dialog` objects.  But maybe we should just make a Dialog object at import time...

Comment (bmw) [inline certbot/display/util.py]: But why not put in `__init__` where we have a `dialog.Dialog` object?

**Comment by bmw** [comment]
Other than moving the assertion up and the logging level, this LGTM.

**Comment by bmw** [comment]
LGTM

**Pull request merged**